src/watch.rs

```diff
@@ -40,84 +45,95 @@ where
             );
 
             let curr_dir = env::current_dir()?;
-            let (_config, _watch_forward, _watch_reverse, _has_reverse) =
+            let (config, mut watch_sources_new, watch_code_new, _has_reverse) =
                 cmd::run_with_args(&args, Some(change == ChangeType::Code))?;
             env::set_current_dir(&curr_dir)?;
+
+            watch_sources_new.insert(config);
+
+            update_watcher(&mut sw, &watch_sources_old, &watch_sources_new)?;
+            update_watcher(&mut cw, &watch_code_old, &watch_code_new)?;
+
+            watch_sources_old = watch_sources_new;
+            watch_code_old = watch_code_new;
         }
     }
 
     Ok(())
 }
 
-/// Calls the closure when a book source file is changed, blocking indefinitely.
-fn trigger_on_change<I, J>(
-    watch_sources: &mut I,
-    watch_code: &mut J,
-) -> Fallible<Receiver<ChangeType>>
-where
-    I: Iterator<Item = PathBuf>,
-    J: Iterator<Item = PathBuf>,
-{
-    use notify::RecursiveMode::*;
+fn update_watcher(
+    watcher: &mut RecommendedWatcher,
+    old_files: &HashSet<PathBuf>,
+    new_files: &HashSet<PathBuf>,
+) -> Fallible {
+    for path in old_files.difference(new_files) {
+        watcher.unwatch(path)?;
+    }
+    for path in new_files.difference(old_files) {
+        watcher.watch(&path, RecursiveMode::NonRecursive)?;
+    }
+    Ok(())
+}
 
+/// Calls the closure when a book source file is changed, blocking indefinitely.
+fn trigger_on_change(
+    watch_sources: HashSet<PathBuf>,
```

Comment:
One consequence of returning instead of the closure should that you do not need ownership here anymore, i.e. `&HashSet<PathBuf`> or even `impl Iterator<Item=PathBuf>` should now suffice. (And hence the cloning be unnecessary.)